Fix Start-PSBuild on WSL if repository was already built on Windows

bergmeister opened the pull request:
To fix #5194: Use the first 3 characters of the passed `-Runtime` parameter for the .inc file (the full runtime name could not be used since the separation characters ike e.g. the the hyphen would make dotnet crash when building).

A separate issue is that the runtime does not get determined if is not being passed (which is OK since dotnet takes then care of choosing the right one) but maybe given this scenario would it make sense to determine and set it as a convenience?

bergmeister committed:
Fix build on WSL if repository was already built on Windows.
Note that the '-Runtime' parameter needs to be passed in this case.

rkeithhill commented:
This should slot in nicely with my tweak to build.psm1 #5341

rkeithhill commented:
I think you need to update this .gitignore file to make it handle the platform specific inc files:
```
40:121ms> gc .\src\TypeCatalogGen\.gitignore
powershell.inc
```
I was noticing that Git thought I had a new file with powershell_win.inc.

bergmeister committed:
Update .gitignore to match powershell**.inc instead of powershell.inc because the file can now optionally end in the runtime prefix.

bergmeister commented:
Thanks. Well spotted, I updated it to ignore now `powershell*.inc`. I find it a bit odd that there is a special `.gitignore` file in this folder and am wondering if we can move this to the top level `.gitignore` file?

andschwa commented:
@bergmeister Because this ignore rule doesn't apply to anywhere else in the repository, just keeping it simple 😄

bergmeister commented:
@andschwa My thinking is rather that if there will never be any other `.inc` files somewhere else, it would be nice to put it into the top one to have a more central place, especially since there is only one entry in it. But this is probably more of a stylistic preference with no 'right' answer depending on how big the central place can become (to me the output of `gci -Recurse .gitignore | % {cat $_}` did not seem to be too long)... But I'll leave it then as it is if that's your preference.

bergmeister commented on build.psm1:
```diff
@@ -489,9 +489,13 @@ Fix steps:
-    if ($TypeGen -or -not (Test-Path "$PSScriptRoot/src/System.Management.Automation/CoreCLR/CorePsTypeCatalog.cs")) {
+    $incFileName = 'powershell'
+    if (-not [string]::IsNullOrEmpty($Runtime)) {
+        $incFileName = "$($incFileName)_$($Runtime.Substring(0,3))" # use the first 3 characters only because otherwise dotnet will crash due to the separation characters
```
You're right about the comment. I noticed that only after the commit but I will fix that. The separation character is only for readability because to me `powershell_win` or `powerhell_lin` is easier and faster to read than `powershellwin` or `powershelllin`

bergmeister committed:
Put comment on single line as per style guide and add more comment details

bergmeister commented:
NB: The build was read because NuGet has a global outage (see [here](https://twitter.com/nuget/status/927630886509539332))

iSazonov commented on build.psm1:
```diff
@@ -489,9 +489,15 @@ Fix steps:
+        # Append the first 3 characters fo the Runtime only because otherwise dotnet will crash due to the separation characters. The underscore is only for better readability.
+        $incFileName = "$($incFileName)_$($Runtime.Substring(0,3))"
+    }
+    if ($TypeGen -or -not (Test-Path "$PSScriptRoot/TypeCatalogGen/$($incFileName).inc")) {
```
We use `$incFileName).inc` so many times so make sense add:
```powershell
$incFileName = ($incFileName).inc
```

bergmeister replied:
Ok, I have addressed that.

iSazonov commented:
I'd prefer to fix the issue more natively like #3870.

Also I'm not sure that's the only runtime addiction. Should we clean up if our files contains "#if UNIX"?

bergmeister committed:
Include file extension in $incFileName variable to reduce '.inc' duplication as suggested in PR.
Tested again that it still works when building on Windows first using 'Start-PSBuild' and then 'Start-PSBuild -Runtime linux-x64' on WSL (ubuntu16)

daxian-dbw commented on build.psm1:
```diff
@@ -489,9 +489,15 @@ Fix steps:
+    if (-not [string]::IsNullOrEmpty($Runtime)) {
+        # File name must be different for Windows and Linux to allow build on Windows and WSL.
+        # Append the first 3 characters fo the Runtime only because otherwise dotnet will crash due to the separation characters. The underscore is only for better readability.
+        $incFileName = "$($incFileName)_$($Runtime.Substring(0,3)).inc"
```
Maybe just use `powershell_$($Runtime.Substring(0,3)).inc`? Otherwise it would beome `powershell.inc_xxx.inc`

bergmeister replied:
This change has now been undone since we do not need that logic any more due to the usage of `$Options.Runtime`

daxian-dbw commented on build.psm1:
```diff
@@ -489,9 +489,15 @@ Fix steps:
+        # Append the first 3 characters fo the Runtime only because otherwise dotnet will crash due to the separation characters. The underscore is only for better readability.
+        $incFileName = "$($incFileName)_$($Runtime.Substring(0,3)).inc"
+    }
+    if ($TypeGen -or -not (Test-Path "$PSScriptRoot/TypeCatalogGen/$($incFileName)")) {
```
`$($incFileName)` can be replaced by `$incFileName`

bergmeister replied:
No, this is one of the gotchas about PowerShell string interpolation because PowerShell allows underscores for variable names. Therefore PowerShell would then look for a variable named `incFileName_` instead of `incFileName`. As a general rule, I always wrap variables in a string inside `$()` because it also safeguards against cases like calling a property on an object...

daxian-dbw replied:
In the case of `"$($incFileName)_xxx.inc"`, you can replace it with `"${incFileName}_xxx.inc"`

daxian-dbw commented on build.psm1:
```diff
@@ -1631,7 +1637,12 @@ function Start-DevPowerShell {
+    (
+        [Parameter(Mandatory=$true)]
+        [ValidateNotNullOrEmpty()]
+        $IncFileName
```
Maybe we should have 'powershell.inc' as the default value for this parameter, so that `Start-TypeGen` without parameter will continue to work.

bergmeister replied:
I can do that if people actually call `Start-TypeGen`. I always use the `-ResGen` swich on `Start-PSBuild` because the reason why I need update ResGen is because I need to build anyway.

daxian-dbw replied:
It's not commonly needed, but when debugging anything about the type catalog, it's handy to just run `Start-TypeGen` without building the rest of the projects. For that purpose, a simple default value like `powershell.inc` would be sufficient, no need to call `New-Options` to figure out the runtime.

daxian-dbw replied:
It's not commonly needed, but when debugging anything about the type catalog, it's handy to just run `Start-TypeGen` without building the rest of the projects. For that purpose, a simple default value like `powershell.inc` would be sufficient, no need to call `New-Options` to figure out the runtime.

daxian-dbw replied:
And a minor comment: the variable name you used in the rest of this function is `incFileName`, maybe it can be changed to `IncFileName`.

daxian-dbw commented on build.psm1:
```diff
@@ -489,9 +489,15 @@ Fix steps:
     # handle TypeGen
-    if ($TypeGen -or -not (Test-Path "$PSScriptRoot/src/System.Management.Automation/CoreCLR/CorePsTypeCatalog.cs")) {
+    $incFileName = 'powershell.inc'
+    if (-not [string]::IsNullOrEmpty($Runtime)) {
```
The `$Options.Runtime` will always be 'NotNullOrEmpty', so maybe you can have:
```
$incFileName = "powershell_$($Runtime.Substring(0,3)).inc"
```

bergmeister replied:
Good idea. Yes, we can use `$Options.Runtime` indeed instead. I was not aware that you already figure it out automatically when it is not being provided. Shall I maybe add this information as a comment to the `$Runtime` parameter?

daxian-dbw replied:
You mean the `-Runtime` parameter of `Start-PSBuild`? Sure you can add a comment there.

daxian-dbw commented on build.psm1:
```diff
@@ -1652,15 +1663,15 @@ function Start-TypeGen
     Push-Location "$PSScriptRoot/src/Microsoft.PowerShell.SDK"
     try {
-        $ps_inc_file = "$PSScriptRoot/src/TypeCatalogGen/powershell.inc"
+        $ps_inc_file = "$PSScriptRoot/src/TypeCatalogGen/$($incFileName)"
```
`$($incFileName)` can be replaced by `$incFileName`

bergmeister replied:
In this case we can do that but what do you think about my comment above where this is not always possible?

daxian-dbw replied:
PowerShell support this syntax for variable: `${var name}`, so you can use `"${varname}_xxx.inc"` in that case.

bergmeister commented:
I'll address the comments next Monday when I'm back from travelling.

bergmeister commented on build.psm1:
```diff
@@ -489,9 +489,15 @@ Fix steps:
+    $incFileName = 'powershell.inc'
+    if (-not [string]::IsNullOrEmpty($Runtime)) {
+        # File name must be different for Windows and Linux to allow build on Windows and WSL.
+        # Append the first 3 characters fo the Runtime only because otherwise dotnet will crash due to the separation characters. The underscore is only for better readability.
```
Thanks for spotting the typo and thanks for coming back about the crash. When I started out I was experimenting with different approaches (one of was to put the inc file into a runtime specific folder) and I actually cannot repro the crash anymore and will therefore use the full `$RunTime` string instead of the first 3 characters.

bergmeister committed:
Fix typo and do not trim the $Runtime string since this does not seem to be necessary.

bergmeister committed:
Use $Options.Runtime instead of $Runtime because the former gets determined if the $Runtime is not being provided.

bergmeister commented:
@daxian-dbw Thanks for your useful feedback. I have addressed all issues and left comments there. But basically the 2 remaining questions are:

- Should we be strict about string interpolation (i.e. use `$()`) since there are gotchas as I pointed out?
- Should `Start-TypeGen` have a default value for the new `$IncFileName` parameter or should we automatically determine it in there similarly as in `Start-PSBuild`?

daxian-dbw commented on build.psm1:
```diff
@@ -488,10 +488,12 @@ Fix steps:
+    # Handle TypeGen
+    # .inc file name must be different for Windows and Linux to allow build on Windows and WSL.
+    $incFileName = "powershell_$($Options.Runtime).inc"
+    if ($TypeGen -or -not (Test-Path "$PSScriptRoot/TypeCatalogGen/$($incFileName)")) {
```
> $($incFileName)

Subexpression is not needed here, you can just use `$incFileName`

daxian-dbw commented:
@bergmeister For the two questions, I answered them in the corresponding comment threads: https://github.com/PowerShell/PowerShell/pull/5346#discussion_r150988197 and https://github.com/PowerShell/PowerShell/pull/5346#discussion_r150988544

bergmeister committed:
Addressed minor suggestions made in PR 5346.
Use default value for Start-TypeGen for convenience and minor syntax changes.
Tested again that it still works when building on Windows first and then on WSL

bergmeister commented:
@daxian-dbw Thanks for the review. I have addressed your comments.

daxian-dbw reviewed (approved):
LGTM. Thanks @bergmeister!
